### ISSUE TITLE
Some prototype functions to recycle contents from existing protocol versions

### ISSUE THREAD

**Issue by florisvdh** [inbo/protocolhelper#14]
@hansvancalster I hope these examples are of help. Does the `system()` command work in Windows? If not we would need to dive in using `git2r`. The below is just a concept that needs further implementation, especially the way of inserting it into the master Rmd as a child. @ThierryO 

You need to have fetched the remote repository (`git fetch origin`) for this example to work, as it requires local availability of the requested tag. It does not matter at which commit/branch you are checked out.

```r
library(dplyr)
library(rprojroot)
library(rmarkdown)

# some objects for testing
protocol_code <- NULL
file_name <- "07_stappenplan.Rmd"
version_number <- NULL

construct_protocolpath <- function(protocol_code, language) { # an internal function
  find_root_file("src/thematic/4_vegetation/sfp-401_korte-titel_nl",
                 criterion = is_git_root) # just an example, to be worked out
}

construct_tag <- function(protocol_code, version_number) { # an internal function
  "example-0.1"  # just an example that exists in the remote.
                 # function is to be worked out
}

recycle_protocolsection <-   # function for the user
  function(protocol_code,
           version_number, 
           language,
           file_name, 
           section = NULL,  # default: take whole file, 
                            # otherwise markdown ref can be put here (default is
                            # lowercase-title-words-with-hyphens)
           demote = 1) {  # demote: number of '#' to prefix to all titles 
                          # before inserting in current protocol
    git_filepath <- 
      construct_protocolpath(protocol_code, language) %>% 
      file.path(file_name) %>% 
      relative_to(find_root(is_git_root), .)
    tag <- construct_tag(protocol_code, version_number)
    gitcommand <- paste0("git show ",
                         tag, ":",
                         git_filepath)
    system(gitcommand,
           intern = TRUE) %>% 
      paste0(collapse = "\n") # further needed before knitting: 
                              # - handling the section and demote arguments
                              # - inserting result into master Rmd
}
```
<details> <summary><strong>My output</strong></summary>

```r
> library(dplyr)

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union


> library(rprojroot)

> library(rmarkdown)

> # some objects for testing
> protocol_code <- NULL

> file_name <- "07_stappenplan.Rmd"

> version_number <- NULL

> construct_protocolpath <- function(protocol_code, language) { # an internal function
+   find_root_file("src/thematic/4_vegetation/sfp-401_korte-tit ..." ... [TRUNCATED] 

> construct_tag <- function(protocol_code, version_number) { # an internal function
+   "example-0.1"  # just an example that exists in the remote.
+  .... [TRUNCATED] 

> git_filepath <- 
+   construct_protocolpath(protocol_code, language) %>% 
+   file.path(file_name) %>% 
+   relative_to(find_root(is_git_root), .)

> tag <- construct_tag(protocol_code, version_number)

> gitcommand <- paste0("git show ",
+                      tag, ":",
+                      git_filepath)

> system(gitcommand,
+        intern = TRUE) %>% 
+   paste0(collapse = "\n")
[1] "# Werkwijze\n\n## Uitvoering\n\n\nGedetailleerde omschrijving van alle stappen die doorlopen moeten worden om het veldprotocol uit te voeren. Subtitels gebruiken om elke stap te omschrijven. \n\n\n### Subtitel 1\n\nUitleg tekst 1\n\n### Subtitel 2\n\nUitleg tekst 2\n\n## Registratie en bewaring van resultaten\n\n\nOpsomming van alle resultaten die bekomen worden na de uitvoering van het protocol en hoe die resultaten geregistreerd, bewaard of opgeslagen moeten worden.\nVoor metingen of observaties: verwijs naar invulformulier (met versienummer; invulformulier toevoegen in bijlage; eventueel bepaalde zaken van invulformulier verduidelijken) en/of naar apparatuur en/of softwareprogramma indien gegevens digitaal worden ingevoerd (verwijs naar SIP indien beschikbaar; indien geen SIP beschikbaar, geef de nodige instructies om gegevens op gepaste wijzen in te kunnen geven).\nVoor staalnames: geef aan hoe de stalen bewaard, gelabeld en vervoerd moeten worden. Geef aan of de stalen voorbehandeld moeten worden. \nVoor digitale foto’s en/of andere digitale bestanden: geef aan hoe en waar deze bestanden moeten worden opgeslagen. Geef eventueel aan welke bestandnamen aan de bestanden moeten gegeven worden.\n"
```
</details>

Fetching the git repo first could be incorporated too as `system("git fetch origin")` or further generalized by:

```r
firstremote <- system("git remote", intern = TRUE)[1] # this usually returns "origin"
system(paste0("git fetch ", firstremote))
```

**Comment by hansvancalster** [comment]
@florisvdh yes! I can confirm that this works on Windows also. So we have a proof of concept :smiley: 
I only noticed encoding trouble in my output, but not yours: fotoâ€™s appeared in my output.

**Comment by hansvancalster** [comment]
Also nice sketch of the user function. This will work well I believe.

**Comment by florisvdh** [comment]
As a sidenote, on Windows it is more robust to use `shell()` (a Windows-only function) instead of `system()`. Some exploration of those commands in Windows (for using GRASS modules) can be found in https://github.com/inbo/n2khab-preprocessing/pull/43#issuecomment-744683452. A more generic custom function `execshell()` is recommendable for better cross-platform use:

```r
execshell <- function(commandstring, intern = FALSE) {
    if (.Platform$OS.type == "windows") {
        res <- shell(commandstring, intern = TRUE)
    } else {
        res <- system(commandstring, intern = TRUE)
    }
    if (!intern) cat(res, sep = "\n") else return(res)
}
```

**Comment by florisvdh** [comment]
Another note. It appears that tidyverse packages shifted from `git2r` to **[`gert`](https://docs.ropensci.org/gert/)** (by Jeroen Ooms) for calling git. From its general aims and purpose, it seems preferable to try using `gert` here, instead of the `git2r` examples above. @hansvancalster

**Comment by hansvancalster** [comment]
I think we are not using `git2r`, just the shell / system functions to send git commands. But I will look into gert.

**Comment by florisvdh** [comment]
Right indeed. There is choice.

**Comment by hansvancalster** [comment]
closed by #37
